### ISSUE TITLE
Pixel Shuffle equivalent

### ISSUE THREAD

**Issue by csolorio** [Microsoft/CNTK#2501]
Hi, 

I'm currently trying to migrate from torch to CNTK and wondered if there's some way to implement the `PixelShuffle` method from Torch (https://github.com/torch/nn/blob/master/PixelShuffle.lua http://pytorch.org/docs/0.1.12/nn.html?highlight=pixel%20shuffle#torch.nn.PixelShuffle) with an already implemented function?

**Comment by ebarsoumMS** [comment]
Hi @csolorio 

We don't have build-in pixel shuffle like API, however, it can be implemented using reshape and transpose, as shown below. The below code is adapted from: https://github.com/pytorch/pytorch/issues/1684 

```
import CNTK as C

def pixel_shuffle(input, scale_factor):
    in_channels, in_height, in_width = input.shape

    out_channels = in_channels // (scale_factor * scale_factor)
    out_height = in_height * scale_factor
    out_width = in_width * scale_factor

    if scale_factor >= 1:
        input_view = C.reshape(input,(out_channels, scale_factor, scale_factor, in_height, in_width))
        shuffle_out = C.transpose(input_view, perm=(0, 3, 1, 4, 2))

    return C.reshape(shuffle_out, (out_channels, out_height, out_width))
```

**Comment by csolorio** [comment]
Wow, thank you so much! Great support :)

**Comment by csolorio** [comment]
During my quest to make the migration of a model, I encountered the following cases in torch that I'm not sure I can reimplement in CNTK: 
1. Dynamically changing the model during training.
```
        if isSwap then
            --Fast model swap
            tempModel = self.model
            self.model = self.util:swapModel(self.model, scaleIdx)
        end

        self.currentErr = 0
	self.model:forward(self.input.train)
	self.currentErr = self.currentErr + self.criterion(self.model.output, self.target)
	self.model:backward(self.input.train, self.criterion.gradInput)

        if isSwap then
            --Return to original model
            self.model = tempModel
        end
```
where ``swapModel`` chooses some branchs from parallel containers and concat tables, like:
```
	local sModel = nn.Sequential()
	sModel
		:add(model:get(1))
		:add(model:get(2))
		:add(model:get(3):get(index))
		:add(model:get(4))
		:add(model:get(5):get(index))
		:add(model:get(6):get(index))
```
Can I use the cntk functions of the same name (forward/backward) or the layer selections for the same purpose?

2.

**Comment by ebarsoumMS** [comment]
Sorry for late reply,  we have sequential and here a consice way to create a model:

```Python
def create_basic_model_with_batch_normalization(input, out_dims):

    with C.layers.default_options(activation=C.relu, init=C.glorot_uniform()):
        model = C.layers.Sequential([
            C.layers.For(range(3), lambda i: [
                C.layers.Convolution((5,5), [32,32,64][i], pad=True),
                C.layers.BatchNormalization(map_rank=1),
                C.layers.MaxPooling((3,3), strides=(2,2))
            ]),
            C.layers.Dense(64),
            C.layers.BatchNormalization(map_rank=1),
            C.layers.Dense(out_dims, activation=None)
        ])

    return model(input)
```
For details, look at https://github.com/Microsoft/CNTK/blob/master/Tutorials/CNTK_201B_CIFAR-10_ImageHandsOn.ipynb

For swapping models,  can you provide more details on the scenario? You might be able to do it using Clone in CNTK.

For (3) yes.

**Comment by spandantiwari** [comment]
New nodes, named `DepthToSpace `and `SpaceToDepth`, which are the equivalent of Pixel Shuffle have been added to master with this commit: https://github.com/Microsoft/CNTK/commit/4bb4dbb5d00756d00823e0a370e56779760cbb68
